server/src/main/java/org/elasticsearch/threadpool/Scheduler.java

```diff
@@ -276,7 +277,11 @@ protected void afterExecute(Runnable r, Throwable t) {
             if (t != null) return;
             // Scheduler only allows Runnable's so we expect no checked exceptions here. If anyone uses submit directly on `this`, we
             // accept the wrapped exception in the output.
-            ExceptionsHelper.reThrowIfNotNull(EsExecutors.rethrowErrors(r));
+            if (r instanceof RunnableFuture && ((RunnableFuture<?>) r).isDone()) {
```

Comment:
I added the assertion in 3781c31015144134a982c0dba42afeaacc549f53, since moving the check to `EsExecutors.rethrowErrors` would mean that we would silently ignore if isDone was false also for non scheduled cases.